### PR TITLE
[211] set and report application name to server 

### PR DESCRIPTION
**Pull request by d-w-moore**
The below example client would report  "myApp" as its **application name** to the iRODS server, and you can see this
via the  `ips` command while the example is running. If the `name` keyword is not supplied for iRODSSession(...), the  contents of the `spOption` environment variable or the default app name of "python-irodsclient" will be assigned, currently with that order of override, though that is subject to change.

```
import time
import os
from irods.session import iRODSSession

if __name__ == '__main__':
    try:
        env_file = os.environ['IRODS_ENVIRONMENT_FILE']
    except KeyError:
        env_file = os.path.expanduser('~/.irods/irods_environment.json')

    with iRODSSession (irods_env_file = env_file, application_name = "myApp") as session:
        session.collections.get('/{0.zone}/home/{0.username}'.format(session))
        time.sleep(15.0)
```

### CODE REVIEW

**Comment by trel** [comment]
Did you mean to include the 156 commit here?       Otherwise, looks good to me.

Still awaiting comment from @kript on whether he would like to have the `spOption` environment variable honored.  I'm still leaning away from its necessity.

**Comment by d-w-moore** [comment]
The [156] commit is technically unrelated, but this morning I discovered its necessity ... should I parcel it out into its own PR thing?

**Comment by d-w-moore** [comment]
If we're leaning toward deprecating `spOption`, perhaps it's better to leave that out.  When we're ready with the final bows (probably still need a test around this ?) and ready to merge, we can remove the environment variable if we haven't heard back.

**Comment by trel** [comment]
the 156 is fine - just seemed to get the streams crossed.

we're not suggesting deprecating `spOption`... we're just considering not extending it's historical use in other places to here.

**Comment by kript** [comment]
Apologies, I had thought the question had been decided rather than moved here.
I don't have a strong steer from my userbase - not many people use the python API outside of sysadmins (although we very nearly had a web app written in it), so I suspect as long as there was a documented way of setting the application name it would be fine. The ability to dynamically change this via environment variable rather than changing code (or having to have the option written into the code) is always of value.

However (and possibly a larger issue than this issue/commit) - the ability to set this context/metadata for clients _at all_ has given us the observability in production to be able to identify and assist people who are using lagging versions, so as there is already a way to do this, I'm in favour of making it _more_ visible, and therefore having a consistent way accross the C and python API seems a good idea to me (as the perosn who isn't going to have to do the work to maintain it...).

**Comment by d-w-moore** [comment]
I see the value of  keeping `spOption` in, esp when launching from a shell script (as systems administrators are prone to do) where -- of course -- environment variables are the simplest way of communicating with the subshell.  No slight is implied there; quite a bit of Bash and Perl haunts my past.

**Comment by trel** [comment]
So, if we keep `spOption` in this PR - I think we should honor the `spOption` over an `application_name` set in the code?

Does that sound correct?

**Comment by d-w-moore** [comment]
spOption as highest precedence sounds correct to me. Environment / config wins over the application's internal settings and defaults, is the pattern of most things in Unix.

File: irods/connection.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ class Connection(object):
 
     DISALLOWING_PAM_PLAINTEXT = True
 
-    def __init__(self, pool, account):
+    def __init__(self, pool, account ):
```
Comment (trel) [inline irods/connection.py:40]: extra space

File: irods/pool.py [hunk 2]
```diff
@@ -1,21 +1,31 @@
...
+
+
 class Pool(object):
 
-    def __init__(self, account):
+    def __init__(self, account, application_name = ''):
+        '''
+        Pool( account , client_name='' )
+        Create an iRODS connection pool; 'account' is an irods.account.iRODSAccount instance
+        and 'client_name' specifies an application name as it should appear in an 'ips' listing.
```
Comment (trel) [inline irods/pool.py:21]: 'client_name' still here.

File: irods/pool.py [hunk 3]
```diff
@@ -1,21 +1,31 @@
...
+        '''
         self.account = account
         self._lock = threading.RLock()
         self.active = set()
         self.idle = set()
         self.connection_timeout = DEFAULT_CONNECTION_TIMEOUT
+        self.application_name = (application_name  or  os.environ.get('spOption','')  or  DEFAULT_APPLICATION_NAME)
```
Comment (trel) [inline irods/pool.py:28]: flip these?   spOption to take precedence.

**Comment by d-w-moore** [comment]
The above are done and the new spOption precedence has been tested. Ready for merge.

**Comment by trel** [comment]
please either remove the 156 commit from this PR or # it

**Comment by d-w-moore** [comment]
Ok, done

**Pull request merged**